desugarer.go

```diff
@@ -16,15 +16,387 @@ limitations under the License.
+		if field.expr1 != nil {
+			err := desugar(&field.expr1, objLevel)
+			if err != nil {
+				return err
+			}
+		}
+		err := desugar(&field.expr2, objLevel+1)
+		if err != nil {
+			return err
+		}
+		if field.expr3 != nil {
+			err := desugar(&field.expr3, objLevel+1)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Simplify asserts
+	// TODO(dcunnin): this
+	for _, field := range *fields {
+		if field.kind != astObjectAssert {
+			continue
+		}
+		/*
+			AST *msg = field.expr3
+			field.expr3 = nil
+			if (msg == nil) {
+				auto msg_str = U"Object assertion failed."
+				msg = alloc->make<LiteralString>(field.expr2->location, msg_str,
+												 LiteralString::DOUBLE, "")
+			}
+
+			// if expr2 then true else error msg
+			field.expr2 = alloc->make<Conditional>(
+				ast->location,
+				field.expr2,
+				alloc->make<LiteralBoolean>(E, true),
+				alloc->make<Error>(msg->location, msg))
+		*/
+	}
+
+	// Remove methods
+	// TODO(dcunnin): this
+	for _, field := range *fields {
+		if !field.methodSugar {
+			continue
+		}
+		/*
+			field.expr2 = alloc->make<Function>(
+				field.expr2->location, field.ids, false, field.expr2)
+			field.methodSugar = false
+			field.ids.clear()
+		*/
+	}
+
+	// Remove object-level locals
+	newFields := []astObjectField{}
+	var binds astLocalBinds
+	for _, local := range *fields {
+		if local.kind != astObjectLocal {
+			continue
+		}
+		binds = append(binds, astLocalBind{variable: *local.id, body: local.expr2})
+	}
+	for _, field := range *fields {
+		if field.kind == astObjectLocal {
+			continue
+		}
+		if len(binds) > 0 {
+			field.expr2 = &astLocal{astNodeBase{loc: *field.expr2.Loc()}, binds, field.expr2}
+		}
+		newFields = append(newFields, field)
+	}
+	*fields = newFields
+
+	// Change all to FIELD_EXPR
+	for i := range *fields {
+		field := &(*fields)[i]
+		switch field.kind {
+		case astObjectAssert:
+		// Nothing to do.
+
+		case astObjectFieldID:
+			field.expr1 = makeStr(string(*field.id))
+			field.kind = astObjectFieldExpr
+
+		case astObjectFieldExpr:
+		// Nothing to do.
+
+		case astObjectFieldStr:
+			// Just set the flag.
+			field.kind = astObjectFieldExpr
+
+		case astObjectLocal:
+			return fmt.Errorf("INTERNAL ERROR: Locals should be removed by now.")
+		}
+	}
+
+	// Remove +:
+	// TODO(dcunnin): this
+	for _, field := range *fields {
+		if !field.superSugar {
+			continue
+		}
+		/*
+			AST *super_f = alloc->make<SuperIndex>(field.expr1->location, field.expr1, nil)
+			field.expr2 = alloc->make<Binary>(ast->location, super_f, BOP_PLUS, field.expr2)
+			field.superSugar = false
+		*/
+	}
+
+	return nil
+}
+
+func desugar(astPtr *astNode, objLevel int) error {
+	ast := *astPtr
+	// TODO(dcunnin): Remove all uses of unimplErr.
+	unimplErr := makeStaticError(fmt.Sprintf("Desugarer does not yet implement ast: %s", reflect.TypeOf(ast)), *ast.Loc())
+
+	var err error // Make all recursive calls of the form err = desugar, rather than some being err := desugar
```

Comment:
I think you can define this variable as part of the return -- 

```
func foo(...) (err error)
```

Then you can either just `return` or still `return err`.